### ISSUE TITLE
Special case implementation of Expr<bool> that reduces to bool.

### ISSUE THREAD

**Issue by arlyon** [cfallin/boolean_expression#12]
Hi!

Love the library. I am currently mapping my `Expr<T>` into `Expr<bool>` and would like to avoid providing a hashmap with true in it. Would it be an idea for something along the lines of the snippet below (naming tbd) to directly return a boolean?

```rust
impl Expr<bool> {
  pub fn evaluate_bool(self) -> bool {
    ...
  }
}
```

This would be cleaner (and maybe slightly faster) than

```rust
expr.map(|f| f.do_stuff()).evaluate(&[(true, true)].iter().cloned().collect())
// vs
expr.map(|f| f.do_stuff()).evaluate_bool()
```

**Comment by cfallin** [comment]
Hi @arlyon,

I'm happy to add convenience APIs like this -- I can review a PR if you'd like to put one together! I might make the name slightly more descriptive: something like `evaluate_with_bool_terminals()` or something of the sort (if I understand correctly).

Just as a tangential thought, I wonder if an API like `evaluate_with<F: Fn(T) -> bool>(&self, f: F) -> bool` would make sense? (The function would provide the truth value for each variable.) This essentially would fold the `map` above into the evaluate and avoid constructing the intermediate `Expr`. It could even be the basis for `evaluate()` -- the hashmap lookup is just one possible closure to plug into the generic. Thoughts?

**Comment by arlyon** [comment]
`evaluate_with` sounds like a great idea.

I was also thinking about something (unrelated) like

```rust
impl<T> From<T> for Expr<T> {
  // Expr::Terminal
}
```

I'll open some PRs and we can continue discussion there. Thanks for the quick reply :)

**Comment by cfallin** [comment]
Sounds good! The `From` impl seems reasonable as well.

**Comment by cfallin** [comment]
This issue is closed with #13, #14 and #15, I think. I've bumped the version to 0.4.2 and published it to crates.io as well.

Thanks again for your contributions!
